hashtable/hashmap.go

```diff
@@ -0,0 +1,161 @@
+package hashtable
+
+const (
+	// KeyNotFound is a constant indicates that
+	// a key cannot be found in the HashMap.
+	KeyNotFound     = 0
+	defaultCapacity = 8
+	upperLoadFactor = 0.75
+	lowerLoadFactor = 0.2
+)
+
+type node struct {
+	key   int
+	value int
+	next  *node
+}
+
+// HashMap is a struct for storing key-value data.
+type HashMap struct {
+	items []*node
+	size  int
+}
+
+// NewHashMap creates a new HashMap instance.
+func NewHashMap() *HashMap {
+	return &HashMap{
+		items: make([]*node, defaultCapacity),
+	}
+}
+
+// Insert adds a new pair of key-value to the HashMap.
+// If the key already exists, it overwrites the key's value.
+func (h *HashMap) Insert(key int, value int) {
+	if h.getLoadFactor() >= upperLoadFactor {
+		h.rehash(true)
+	}
+
+	newNode := &node{
+		key:   key,
+		value: value,
+	}
+
+	h.size++
+	index := h.hash(key)
+	curNode := h.items[index]
+
+	if curNode == nil {
+		// first node in the linked list
+		h.items[index] = newNode
+		return
+	}
+
+	for true {
+		if curNode.key == key {
+			// found key to overwrite value
+			curNode.value = value
+			return
+		}
+
+		if curNode.next == nil {
+			// or reach the last node
+			break
+		}
+
+		curNode = curNode.next
+	}
+
+	curNode.next = newNode
+}
+
+// Search returns the value for the given key.
+// If the key cannot be found, it returns KeyNotFound.
+func (h *HashMap) Search(key int) int {
+	index := h.hash(key)
+	curNode := h.items[index]
+
+	if curNode == nil {
+		return KeyNotFound
+	}
+
+	for curNode.key != key {
+		if curNode.next == nil {
+			// reach the last node
+			return KeyNotFound
+		}
+
+		curNode = curNode.next
+	}
+
+	return curNode.value
```

Comment:
Looks like the code is duplicated. There are some pieces of code which follow same steps:
- Check if node is not nil
- Iterate nodes
- Do something with the last node

Could you do it without making duplicated code?